### DISCUSSION TITLE
Assignment of tuple to element in list[tuple] fails

### DISCUSSION THREAD

**Discussion by mikepqr** [microsoft/pyright#4305] [Q&A] [answered]
Line 2 of
```
x = [(0, 0)]
x[0] = (1, 1)
```
passes type-checking in mypy but results in this in pyright:
```
1. Argument of type "Literal[0]" cannot be assigned to parameter "__s" of type "slice" in function "__setitem__"
     "Literal[0]" is incompatible with "slice"
2. Argument of type "tuple[Literal[1], Literal[1]]" cannot be assigned to parameter "__o" of type "Iterable[tuple[Literal[0], Literal[0]]]" in function "__setitem__"
     "tuple[Literal[1], Literal[1]]" is incompatible with "Iterable[tuple[Literal[0], Literal[0]]]"
       TypeVar "_T_co@Iterable" is covariant
         "Literal[1]" is incompatible with "tuple[Literal[0], Literal[0]]"
```
I assume this is intended behavior (seems impossible that I'd be the first person to encounter a bug in such simple code), but I'm struggling to understand this. Can someone help me figure out why this fails? And is there anything I can do to make it pass without changing the semantics of my code?

**Comment by erictraut** [comment, marked as answer]
Yes, this is intended behavior. It's due to the [type inference rules](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#tuple-expressions) pyright uses for tuples. These differ somewhat from the type inference behaviors of mypy. There are pros/cons of each approach, but we believe that pyright's provide better results (fewer false positives) most of the time. Your example is one where the opposite is true.

You can override the type inference behavior by providing a type declaration for the variable. Depending on the intent of your code, one of the following annotations is probably what you're looking for.

```python
# Each tuple in the list must have two entries, each of them an int
x: list[tuple[int, int]]` = [(0, 0)]
```

or 

```python
# Each tuple in the list must have zero or more entries, each of them an int
x: list[tuple[int, ...]]` = [(0, 0)]
```

**Reply by mikepqr** [reply]
Perfect. Thank you so much for the quick reply!
